gdal/frmts/wcs/wcsutils.cpp

```diff
@@ -403,52 +418,95 @@ bool DeleteEntryFromCache(const CPLString &cache_dir, const CPLString &key, cons
+CPLErr SearchCache(const CPLString &cache,
+                   const CPLString &url,
+                   CPLString &filename,
+                   const CPLString &ext,
+                   bool &found)
 {
     found = false;
-    CPLString db_name = CPLFormFilename(cache_dir, "db", NULL);
-    VSILFILE *db = VSIFOpenL(db_name, "r");
-    if (db) {
-        while (const char *line = CPLReadLineL(db)) {
-            char *value = strchr((char *)line, '=');
-            if (*value == '=') {
-                *value = '\0';
-            } else {
-                continue;
-            }
-            if (strcmp(url, value + 1) == 0) {
-                filename = line;
-                found = true;
-                break;
-            }
+    CPLString db = CPLFormFilename(cache, "db", NULL);
+    VSILFILE *f = VSIFOpenL(db, "r");
+    if (!f) {
+        CPLError(CE_Failure, CPLE_FileIO, "Can't open file '%s': %i\n", db.c_str(), errno);
+        return CE_Failure;
+    }
+    while (const char *line = CPLReadLineL(f)) {
+        char *value = strchr((char *)line, '=');
+        if (*value == '=') {
+            *value = '\0';
+        } else {
+            continue;
         }
-        VSIFCloseL(db);
+        if (strcmp(url, value + 1) == 0) {
+            filename = line;
+            found = true;
+            break;
+        }
+    }
+    VSIFCloseL(f);
+    if (found) {
+        filename = CPLFormFilename(cache, (filename + ext).c_str(), NULL);
+        found = FileIsReadable(filename);
+        // if not readable, we should delete the entry
     }
-    if (!found) {
-        db = VSIFOpenL(db_name, "a");
-        if (!db) {
-            CPLError(CE_Failure, CPLE_FileIO, "Can't open file '%s': %i\n", db_name.c_str(), errno);
-            return false;
+    return CE_None;
+}
+
+/* -------------------------------------------------------------------- */
+/*      AddEntryToCache                                                 */
+/*      A new unique key is created into the database based on the      */
+/*      filename replacing X with a random ascii character.             */
+/*      The returned filename is a path formed from the cache directory */
+/*      name, the filename, and the ext.                                */
+/* -------------------------------------------------------------------- */
+
+CPLErr AddEntryToCache(const CPLString &cache,
+                       const CPLString &url,
+                       CPLString &filename,
+                       const CPLString &ext)
+{
+    // assuming the cache was locked to us and the url is not in the cache
+    CPLString store = filename;
+    CPLString db = CPLFormFilename(cache, "db", NULL);
+    VSILFILE *f = VSIFOpenL(db, "a");
+    if (!f) {
+        CPLError(CE_Failure, CPLE_FileIO, "Can't open file '%s': %i\n", db.c_str(), errno);
+        return CE_Failure;
+    }
+        
+    // create a new file into the cache using filename as template
```

Comment:
instead of using randomness, why not using a hash on the URL ? See https://github.com/OSGeo/gdal/blob/trunk/gdal/ogr/ogrsf_frmts/gmlas/ogrgmlasxsdcache.cpp#L124